src/models/aae.jl

```diff
@@ -0,0 +1,282 @@
+	
+	# construct the model
+	# encoder - diagonal covariance
+	encoder_map = Chain(
+		build_mlp(idim, hdim, hdim, nlayers-1, activation=activation)...,
+		ConditionalDists.SplitLayer(hdim, [zdim, zdim], [identity, safe_softplus])
+		)
+	encoder = ConditionalMvNormal(encoder_map)
+	
+	# decoder - we will optimize only a shared scalar variance for all dimensions
+	decoder_map = Chain(
+		build_mlp(zdim, hdim, hdim, nlayers-1, activation=activation)...,
+		ConditionalDists.SplitLayer(hdim, [idim, 1], [identity, safe_softplus])
+		)
+	decoder = ConditionalMvNormal(decoder_map)
+
+	# discriminator
+	discriminator = build_mlp(zdim, hdim, 1, nlayers, activation=activation, lastlayer="σ")
+
+	# prior
+	if prior == "normal"
+		prior_arg = zdim
+	elseif prior == "vamp"
+		(pseudoinput_mean == nothing) ? error("if `prior=vamp`, supply pseudoinput array") : nothing
+		prior_arg = init_vamp(pseudoinput_mean, k)
+	end
+
+	# reset seed
+	(init_seed != nothing) ? Random.seed!() : nothing
+
+	# get the vanilla VAE
+	model = AAE(prior_arg, encoder, decoder, discriminator)
+end
+
+"""
+    dloss(d,g,x,z)
+
+Discriminator loss.
+"""
+dloss(d,g,x,z) = - 0.5f0*(mean(log.(d(x) .+ eps(Float32))) + mean(log.(1 .- d(g(z)) .+ eps(Float32))))
+
+"""
+    gloss(d,g,x)
+
+Generator loss.
+"""
+gloss(d,g,x) = - mean(log.(d(g(x)) .+ eps(Float32)))
+
+"""
+	aeloss(AAE, x[, batchsize])
+
+Autoencoding loss.
+"""
+aeloss(m::AAE,x) = - mean(logpdf(m.decoder, x, rand(m.encoder, x)))
+aeloss(m::AAE,x,batchsize::Int) = 
+	mean(map(y->aeloss(m,y), Flux.Data.DataLoader(x, batchsize=batchsize)))
+
+"""
+	dloss(AAE,x[, batchsize])
+
+Discriminator loss given original sample x.
+"""
+dloss(m::AAE,x) = dloss(m.discriminator, y->mean(m.encoder,y), rand(m.prior, size(x,ndims(x))), x)
+dloss(m::AAE{<:VAMP},x) = dloss(m.discriminator, y->mean(m.encoder,y), mean(m.encoder, rand(m.prior, size(x,ndims(x)))), x)
+dloss(m::AAE,x,batchsize::Int) = 
+	mean(map(y->dloss(m,y), Flux.Data.DataLoader(x, batchsize=batchsize)))
+# note that X and Z is swapped here from the normal notation
+
+"""
+	gloss(AAE,x[, batchsize])
+
+Encoder/generator loss.
+"""
+gloss(m::AAE,x) = gloss(m.discriminator, y->mean(m.encoder,y), x)
+gloss(m::AAE,x,batchsize::Int) = 
+	mean(map(y->gloss(m,y), Flux.Data.DataLoader(x, batchsize=batchsize)))
+
+"""
+	loss(m::AAE, x)
+
+A loss that is used for stopping the training.
+"""
+loss(m::AAE,x) = aeloss(m,x)
+loss(m::AAE,x,batchsize::Int) = aeloss(m,x,batchsize)
+#loss(m::AAE,x) = aeloss(m,x) + dloss(m,x) + gloss(m,x)
+#loss(m::AAE,x,batchsize::Int) = aeloss(m,x,batchsize) + dloss(m,x,batchsize) + gloss(m,x,batchsize)
+
+"""
+	StatsBase.fit!(model::AAE, data::Tuple; max_train_time=82800, 
+	lr=0.001, batchsize=64, patience=30, check_interval::Int=10, kwargs...)
+"""
+function StatsBase.fit!(model::AAE, data::Tuple; max_train_time=82800, lr=0.001, 
+	batchsize=64, patience=30, check_interval::Int=10, kwargs...)
+	history = MVHistory()
+	aeopt = ADAM(lr)
```

Comment:
Why do you have more than one optimizer, if they just share the same learning rate?

Reply:
Because each optimizer optimizes different parts of the model and different set of parameters, therefore the history (moments etc.) is different. Or can this be achieved by using a single optimizer for all parameters?

Reply:
Don't know. 
If it works, don't change it. :smile:

Reply:
originally, I only had one optimizer and was surprised that it worked at all - maybe it is restarted everytime it encounters new parameters? But the convergence is much better with separate optimizers.